pyop2/codegen/rep2loopy.py

```diff
@@ -160,33 +160,10 @@ class INVCallable(LACallable):
     """
     def generate_preambles(self, target):
         assert isinstance(target, loopy.CTarget)
-        inverse_preamble = """
-            #define Inverse_HPP
-            #define BUF_SIZE 30
-
-            static PetscBLASInt ipiv_buffer[BUF_SIZE];
-            static PetscScalar work_buffer[BUF_SIZE*BUF_SIZE];
-            static void inverse(PetscScalar* __restrict__ Aout, const PetscScalar* __restrict__ A, PetscBLASInt N)
-            {
-                PetscBLASInt info;
-                PetscBLASInt *ipiv = N <= BUF_SIZE ? ipiv_buffer : malloc(N*sizeof(*ipiv));
-                PetscScalar *Awork = N <= BUF_SIZE ? work_buffer : malloc(N*N*sizeof(*Awork));
-                memcpy(Aout, A, N*N*sizeof(PetscScalar));
-                LAPACKgetrf_(&N, &N, Aout, &N, ipiv, &info);
-                if(info == 0){
-                    LAPACKgetri_(&N, Aout, &N, ipiv, Awork, &N, &info);
-                }
-                if(info != 0){
-                    fprintf(stderr, \"Getri throws nonzero info.\");
-                    abort();
-                }
-                if ( N > BUF_SIZE ) {
-                    free(Awork);
-                    free(ipiv);
-                }
-            }
-        """
-        yield ("inverse", "#include <petscsys.h>\n#include <petscblaslapack.h>\n" + inverse_preamble)
+        import os
+        with open(os.path.dirname(__file__)+"/c/inverse.c", "r") as myfile:
```

Comment:
Can you open and read these files once (on import on in a prepopulation phase) rather than every time. Additionally, we probably want to read on only one rank on any given communicator (this might require refactoring so that the codegen builder knows about communicators).

Reply:
> Can you open and read these files once (on import on in a prepopulation phase) rather than every time. 
Like that https://github.com/OP2/PyOP2/pull/597/commits/da1de082ebbd25a3598d8bed59ed6bbebd015cdf ?

Reply:
> Additionally, we probably want to read on only one rank on any given communicator (this might require refactoring so that the codegen builder knows about communicators).

I think I will need help with that/more information how exactly to achieve that. Is there are a similar example somewhere?

Reply:
If you have a communicator, then you can do:

```
if comm.rank == 0:
    with open(...) as f:
        source = f.read()
else:
    source = None
source = comm.bcast(source, root=0)
```

Reply:
Okay, and what do you mean with refactoring?

Can we not just do the following?
```
from pyop2.mpi import COMM_WORLD
if COMM_WORLD.rank == 0:
    with open(...) as f:
        source = f.read()
else:
    source = None
source = COMM_WORLD.bcast(source, root=0)
```